src/nupic/support/group_by.py

```diff
@@ -0,0 +1,86 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from itertools import groupby
+
+''' A utility function wrapper around groupby in itertools. Allows to
+    walk across n sorted lists with respect to their key functions
+    and yields a tuple of n lists of the members of the next *smallest*
+    group.
+
+    @param args (list) a list of arguments alternating between sorted lists and
+                       their respective key functions. The lists should be
+                       sorted with respect to their key function.
+
+    @return (tuple) a n + 1 dimensional tuple, where the first element is the
+                    key of the group and the other n entries are lists of
+                    objects that are a member of the current group that is being
+                    iterated over in the nth list passed in.
+
+    Notes: Read up on groupby here:
+           https://docs.python.org/dev/library/itertools.html#itertools.groupby
+'''
+
+def groupByN(*args):
+  groupsList = [] # list of each list's (k, group) tuples
+  indexList = [] # list of [currentIndex, endIndex] pairs
+
+  if len(args) % 2 == 1:
+    raise ValueError("Must have a key function for every list.")
+
+  # zip up tuples of lists and their functions
+  tups = [(args[i], args[i+1]) for i in xrange(0, len(args), 2)]
+
+  # populate above lists
+  for listn, fn in tups:
+    groupsListEntry = [(k, list(g)) for k, g in groupby(listn, fn)]
```

Comment:
Is the `list()` necessary? It seems like we could return a tuple of generators that we got from `groupby`. That would avoid creating a bunch of small lists.

When the list is empty (e.g. when there are no active segments in the column) we could return some other special value like `None`. Then callers don't need to know the length of the list, they can just check for `None` (so we don't have to return counts). This wouldn't be inconsistent with `groupby` usage since there's never a case in groupby where the generator is empty.